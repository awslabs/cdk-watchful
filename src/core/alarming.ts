import { Alarm, AlarmProps, ComparisonOperator, HorizontalAnnotation, IAlarm, IMetric, TreatMissingData } from '@aws-cdk/aws-cloudwatch';
import { Construct } from '@aws-cdk/core';

/**
 * Single alarm customization. If any property is undefined, a default value will be used by the factory.
 */
interface AlarmCustomization {
  /**
   * number of datapoints to alarm on
   * @default factory default
   */
  readonly datapointsToAlarm?: number;
  /**
   * number of periods to evaluate
   * @default factory default
   */
  readonly evaluationPeriods?: number;
  /**
   * whether to enable alarm actions
   * @default factory default
   */
  readonly actionsEnabled?: boolean;
}

/**
 * Single alarm definition with customization.
 */
export interface AlarmDefinition extends AlarmCustomization {
  /**
   * alarm threshold
   */
  readonly threshold: number;
}

/**
 * Properties related to an existing alarm.
 */
export interface CreatedAlarm {
  /**
   * created alarm
   */
  readonly alarm: IAlarm;
  /**
   * annotation of the created alarm
   */
  readonly annotation: HorizontalAnnotation;
  /**
   * the original alarm definition
   */
  readonly definition: AlarmDefinition;
}

/**
 * Factory that creates alarms on metrics.
 */
export interface AlarmFactory {
  /**
   * Creates alarms according to the alarm definitions.
   * @param id alarm identifier (to distinguish alarm from other alarms on the same resource)
   * @param metric metric to alarm on
   * @param op comparison operator
   * @param definitions map of (disambiguator, alarm definition), might be empty or undefined
   */
  createAlarms(id: string, metric: IMetric, op: ComparisonOperator, definitions?: Record<string, AlarmDefinition>): CreatedAlarm[];
}

/**
 * Props to create DefaultAlarmFactory.
 */
export interface DefaultAlarmFactoryProps {
  /**
   * global prefix to be used to all alarms created by this factory
   */
  readonly prefix: string;
  /**
   * default evaluation periods used when there is no custom one
   */
  readonly evaluationPeriodsDefault: number;
  /**
   * default number of datapoints used when there is no custom one
   */
  readonly datapointsToAlarmDefault: number;
  /**
   * default action enabled flag used when there is no custom one
   */
  readonly actionsEnabledDefault: boolean;
}

/**
 * Default implementation of the alarm factory.
 *
 */
export class DefaultAlarmFactory extends Construct implements AlarmFactory {
  protected readonly prefix: string;
  protected readonly evaluationPeriodsDefault: number;
  protected readonly datapointsToAlarmDefault: number;
  protected readonly actionsEnabledDefault: boolean;

  constructor(scope: Construct, id: string, props: DefaultAlarmFactoryProps) {
    super(scope, id);
    this.prefix = props.prefix;
    this.evaluationPeriodsDefault = props.evaluationPeriodsDefault;
    this.datapointsToAlarmDefault = props.datapointsToAlarmDefault;
    this.actionsEnabledDefault = props.actionsEnabledDefault;
  }

  createAlarms(id: string, metric: IMetric, op: ComparisonOperator, definitions?: Record<string, AlarmDefinition>) {
    const createdAlarms: CreatedAlarm[] = [];
    for (let [disambiguator, definition] of Object.entries(definitions ?? {})) {
      createdAlarms.push(this.createAlarm(id, metric, op, disambiguator, definition));
    }
    return createdAlarms;
  }

  protected createAlarm(id: string, metric: IMetric, op: ComparisonOperator, disambiguator: string, props: AlarmDefinition) {
    const alarmId = this.alarmId(id, metric, op, disambiguator, props);
    const alarmName = this.alarmName(id, metric, op, disambiguator, props);
    const alarmDescription = this.alarmDescription(id, metric, op, disambiguator, props);

    const definition: AlarmProps = {
      metric,
      threshold: props.threshold,
      comparisonOperator: op,
      treatMissingData: TreatMissingData.MISSING,
      alarmName,
      alarmDescription,
      evaluationPeriods: props.evaluationPeriods ?? this.evaluationPeriodsDefault,
      datapointsToAlarm: props.datapointsToAlarm ?? this.datapointsToAlarmDefault,
      actionsEnabled: props.actionsEnabled ?? this.actionsEnabledDefault,
    };

    const alarm = new Alarm(this, alarmId, definition);
    const annotation = alarm.toAnnotation();

    return { alarm, definition, annotation };
  }

  protected alarmId(id: string, _metric: IMetric, _op: ComparisonOperator, disambiguator: string, _props?: AlarmDefinition) {
    // TODO: will be improved upon later
    return `${this.prefix}_${id}_${disambiguator}`;
  }

  protected alarmName(_id: string, _metric: IMetric, _op: ComparisonOperator, _disambiguator: string, _props?: AlarmDefinition) {
    // TODO: will be improved upon later
    return undefined;
  }

  protected alarmDescription(_id: string, _metric: IMetric, _op: ComparisonOperator, _disambiguator: string, _props?: AlarmDefinition) {
    // TODO: will be improved upon later
    return undefined;
  }
}
