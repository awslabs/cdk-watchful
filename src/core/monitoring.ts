import { IWidget } from '@aws-cdk/aws-cloudwatch/lib/widget';

/**
 * Represents a monitoring of a single resource or group of related resources.
 * Provides metrics and alarms, and can be subclassed to customize look and feel.
 */
export interface WatchfulMonitoring {
  /**
   * Returns the monitoring widgets.
   * @return widgets
   */
  widgets(): IWidget[];
}
