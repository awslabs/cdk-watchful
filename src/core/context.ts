/**
 * Provides access to global settings and parameters.
 * TODO: currently just a draft
 */
import { Duration } from '@aws-cdk/core';

export class WatchfulContext {
  readonly awsAccountRegion: string;
  readonly defaultMetricPeriod: Duration;

  constructor() {
    // TODO
    this.awsAccountRegion = 'us-east-1';
    this.defaultMetricPeriod = Duration.minutes(5);
  }
}
