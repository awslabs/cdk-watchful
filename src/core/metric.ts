import { DimensionHash, IMetric, MathExpression, Metric } from '@aws-cdk/aws-cloudwatch';
import { Duration } from '@aws-cdk/core';

/**
 * Similar as CloudWatch Statistic, but also with missing percentiles.
 */
export enum WatchfulMetricStatistic {
  SAMPLE_COUNT = 'SampleCount',
  AVERAGE = 'Average',
  SUM = 'Sum',
  MINIMUM = 'Minimum',
  MAXIMUM = 'Maximum',
  P50 = 'p50',
  P90 = 'p90',
  P99 = 'p99',
}

/**
 * Abstract base class for all Watchful metrics.
 */
export abstract class WatchfulMetric {
  /**
   * Converts the object into the standard CloudWatch metric object.
   */
  abstract unwrap(): IMetric;
}

/**
 * Abstract base class for all metrics that support alarming on them.
 */
export abstract class WatchfulMetricWithAlarmSupport extends WatchfulMetric {
  // empty
}

/**
 * Props to create SimpleWatchfulMetric.
 */
export interface SimpleWatchfulMetricProps {
  readonly namespace: string;
  readonly metricName: string;
  readonly statistic: WatchfulMetricStatistic;
  readonly dimensions?: DimensionHash;
  readonly period: Duration;
  readonly label?: string;
  readonly color?: string;
}

/**
 * Simple metric, corresponding with the emitted datapoints.
 */
export class SimpleWatchfulMetric extends WatchfulMetricWithAlarmSupport {
  protected readonly props: SimpleWatchfulMetricProps;

  constructor(props: SimpleWatchfulMetricProps) {
    super();
    this.props = props;
  }

  unwrap(): IMetric {
    return new Metric({
      namespace: this.props.namespace,
      metricName: this.props.metricName,
      dimensions: this.props.dimensions,
      period: this.props.period,
      statistic: this.props.statistic,
      label: this.props.label,
      color: this.props.color,
    });
  }
}

/**
 * Props to create WatchfulMetricMath.
 */
export interface WatchfulMetricMathProps {
  readonly expression: string;
  readonly usingMetrics: Record<string, WatchfulMetric>;
  readonly period: Duration;
  readonly label?: string;
  readonly color?: string;
}

/**
 * Math expression, that transforms different metrics using an expression.
 * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html#metric-math-syntax
 */
export class WatchfulMetricMath extends WatchfulMetricWithAlarmSupport {
  protected props: WatchfulMetricMathProps;

  constructor(props: WatchfulMetricMathProps) {
    super();
    this.props = props;
  }

  unwrap(): IMetric {
    const usingMetrics: Record<string, IMetric> = {};

    for (const [key, value] of Object.entries(this.props.usingMetrics)) {
      usingMetrics[key] = value.unwrap();
    }

    return new MathExpression({
      expression: this.props.expression,
      label: this.props.label,
      color: this.props.color,
      period: this.props.period,
      usingMetrics,
    });
  }
}

/**
 * Props to create WatchfulMetricSearch.
 */
export interface WatchfulMetricSearchProps {
  readonly namespace: string;
  readonly dimensions: Record<string, string>;
  readonly searchQuery?: string;
  readonly statistic: WatchfulMetricStatistic;
  readonly period: Duration;
  readonly label: string;
}

/**
 * Search metric results. Does not support creating any kind of alarms.
 * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/search-expression-syntax.html
 */
export class WatchfulMetricSearch extends WatchfulMetric {
  protected props: WatchfulMetricSearchProps;

  constructor(props: WatchfulMetricSearchProps) {
    super();
    this.props = props;
  }

  unwrap(): IMetric {
    const period = this.props.period;
    const dimensionKeys = Object.keys(this.props.dimensions).join(',');
    const schema = `{${this.props.namespace},${dimensionKeys}}`;

    function mapper(keyAndValue: string[]) {
      const [key, value] = keyAndValue;
      if (value === '') {
        return key;
      } else {
        return `${key}="${value}"`;
      }
    }

    const dimensionKeysAndValues = Object
      .entries(this.props.dimensions)
      .map(mapper)
      .join(' ');

    const expression = `SEARCH('${schema} ${dimensionKeysAndValues} ${this.props.searchQuery ?? ''}', '${this.props.statistic}', ${period.toSeconds()})`;

    return new MathExpression({
      expression,
      usingMetrics: {},
      label: ' ',
      period,
    });
  }
}
