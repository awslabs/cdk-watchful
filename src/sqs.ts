import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as sqs from '@aws-cdk/aws-sqs';
import * as cdk from '@aws-cdk/core';
import { IWatchful } from './api';

export interface WatchSqsOptions {
}

export interface WatchSqsServiceProps extends WatchSqsOptions {
  readonly title: string;
  readonly watchful: IWatchful;
  readonly sqs: sqs.IQueue;
}

export class WatchSqsService extends cdk.Construct {
  private readonly watchful: IWatchful;
  private readonly sqs: sqs.IQueue;

  private messageMetric!: cloudwatch.Metric;
  private messageAlarm!: cloudwatch.Alarm;

  constructor(scope: cdk.Construct, id: string, props: WatchSqsServiceProps) {
    super(scope, id);

    this.watchful = props.watchful;
    this.sqs = props.sqs;

    this.watchful.addSection(props.title, {
      links: [{ title: 'SQS Console', url: linkForSqsService(this.sqs) }],
    });

    this.createSqsMessageMonitor();

    // create widgets
    const messageAlarmWidget = new cloudwatch.AlarmWidget({
      title: 'Messages',
      alarm: this.messageAlarm!,
      width: 12, // half width
    });

    this.watchful.addWidgets(messageAlarmWidget);

  } // constructor
  // helper functions for creating metrics
  private createSqsMessageMonitor() {
    this.messageMetric = new cloudwatch.Metric({
      metricName: SqsMetric.ApproximateNumberOfMessagesVisible,
      namespace: 'AWS/SQS',
      statistic: 'Maximum',
      dimensions: {
        QueueName: this.sqs.queueName,
      },
    });
    this.messageAlarm = new cloudwatch.Alarm(this, 'deliveryToRedshiftAlarm', {
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
      metric: this.messageMetric,
      threshold: 0, // TODO pass parameter.  Used for DLQs only right now
      period: cdk.Duration.minutes(2),
      evaluationPeriods: 1,
      treatMissingData: cloudwatch.TreatMissingData.MISSING,
    });
    this.watchful.addAlarm(this.messageAlarm, false);
  }
}

// TODO extend to monitor other SQS metrics
const enum SqsMetric {
  NumberOfMessagesReceived = 'NumberOfMessagesReceived',
  ApproximateNumberOfMessagesVisible = 'ApproximateNumberOfMessagesVisible',
}

// eslint-disable-next-line @typescript-eslint/no-shadow
function linkForSqsService(sqs: sqs.IQueue) {
  return `https://console.aws.amazon.com/sqs/v2/home?region=${sqs.stack.region}#/queues/https%3A%2F%2Fsqs.${sqs.stack.region}.amazonaws.com%2F${sqs.stack.account}%2F${sqs.queueName}`;
}
