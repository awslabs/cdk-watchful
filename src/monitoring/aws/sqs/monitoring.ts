import { ComparisonOperator, GraphWidget, IWidget, Metric } from '@aws-cdk/aws-cloudwatch';
import { IQueue } from '@aws-cdk/aws-sqs';
import { Stack } from '@aws-cdk/core';
import { AlarmDefinition, CreatedAlarm } from '../../../core/alarming';
import { MonitoringWithAlarms, MonitoringWithAlarmsProps } from '../../../core/monitoring';
import { CountAxis, SizeBytesAxis, TimeSecondsAxis } from '../../../widget/axis';
import { CommonWidgetDimensions } from '../../../widget/constant';
import { SectionWidget, SectionWidgetProps } from '../../../widget/section';
import { SqsMetricFactory } from './metrics';

/**
 * Alarm configuration for an SQS Queue.
 */
export interface SqsMonitoringAlarms {
  /**
   * alarm when queue size drops below threshold
   * @default no alarms
   */
  readonly alarmOnQueueSizeLow?: Record<string, AlarmDefinition>;
  /**
   * alarm when queue size raises above threshold
   * @default no alarms
   */
  readonly alarmOnQueueSizeHigh?: Record<string, AlarmDefinition>;
  /**
   * alarm when max message age in the queue raises above threshold
   * @default no alarms
   */
  readonly alarmOnQueueOldestMessageAgeHigh?: Record<string, AlarmDefinition>;
}

/**
 * Properties to create SqsMonitoring.
 */
export interface SqsMonitoringProps extends MonitoringWithAlarmsProps {
  /**
   * queue to be monitored
   */
  readonly queue: IQueue;
  /**
   * alarm configuration
   * @default no alarms
   */
  readonly alarms?: SqsMonitoringAlarms;
}

/**
 * Monitoring of SQS Queue.
 */
export class SqsMonitoring extends MonitoringWithAlarms {
  protected readonly section: SectionWidgetProps;
  protected readonly metrics: SqsMetricFactory;

  protected readonly visibleMessagesMetric: Metric;
  protected readonly incomingMessagesMetric: Metric;
  protected readonly oldestMessageAgeMetric: Metric;
  protected readonly messageSizeMetric: Metric;

  protected readonly lowCountAlarms: CreatedAlarm[];
  protected readonly highCountAlarms: CreatedAlarm[];
  protected readonly ageAlarms: CreatedAlarm[];

  constructor(props: SqsMonitoringProps) {
    super(props);

    this.section = this.headerProps(props);
    this.metrics = new SqsMetricFactory();

    this.visibleMessagesMetric = this.metrics.metricApproximateVisibleMessages(props.queue.queueName);
    this.incomingMessagesMetric = this.metrics.metricIncomingMessages(props.queue.queueName);
    this.oldestMessageAgeMetric = this.metrics.metricAgeOfOldestMessageInSeconds(props.queue.queueName);
    this.messageSizeMetric = this.metrics.metricAverageMessageSizeInBytes(props.queue.queueName);

    this.lowCountAlarms = this.alarmFactory.createAlarms(
      'QueueSizeLow',
      this.visibleMessagesMetric,
      ComparisonOperator.LESS_THAN_THRESHOLD,
      props.alarms?.alarmOnQueueSizeLow,
    );
    this.highCountAlarms = this.alarmFactory.createAlarms(
      'QueueSizeHigh',
      this.visibleMessagesMetric,
      ComparisonOperator.GREATER_THAN_THRESHOLD,
      props.alarms?.alarmOnQueueSizeHigh,
    );
    this.ageAlarms = this.alarmFactory.createAlarms(
      'QueueOldestMessageAgeHigh',
      this.oldestMessageAgeMetric,
      ComparisonOperator.GREATER_THAN_THRESHOLD,
      props.alarms?.alarmOnQueueOldestMessageAgeHigh,
    );
  }

  getWidgets(): IWidget[] {
    return [
      this.headerWidget(),
      this.messageCountWidget(CommonWidgetDimensions.ThirdWidth, CommonWidgetDimensions.DefaultHeight),
      this.messageAgeWidget(CommonWidgetDimensions.ThirdWidth, CommonWidgetDimensions.DefaultHeight),
      this.messageSizeWidget(CommonWidgetDimensions.ThirdWidth, CommonWidgetDimensions.DefaultHeight),
    ];
  }

  protected headerWidget() {
    return new SectionWidget(this.section);
  }

  protected messageCountWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: 'Message Count',
      left: [this.visibleMessagesMetric, this.incomingMessagesMetric],
      leftYAxis: CountAxis,
      leftAnnotations: [...this.lowCountAlarms.map(a => a.annotation), ...this.highCountAlarms.map(a => a.annotation)],
    });
  }

  protected messageAgeWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: 'Oldest Message Age',
      left: [this.oldestMessageAgeMetric],
      leftYAxis: TimeSecondsAxis,
      leftAnnotations: [...this.ageAlarms.map(a => a.annotation)],
    });
  }

  protected messageSizeWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: 'Message Size',
      left: [this.messageSizeMetric],
      leftYAxis: SizeBytesAxis,
    });
  }

  protected headerProps(props: SqsMonitoringProps) {
    return {
      titleMarkdown: `SQS Queue **${props.queue.queueName}**`,
      quicklinks: this.headerQuickLinks(props),
    };
  }

  protected headerQuickLinks(props: SqsMonitoringProps) {
    return [
      { title: 'Overview', url: this.urlForQueueOverview(props) },
    ];
  }

  protected urlForQueueOverview(props: SqsMonitoringProps) {
    const region = Stack.of(props.queue).region;
    const queueUrl = props.queue.queueUrl;
    return `https://${region}.console.aws.amazon.com/sqs/v2/home?region=${region}#/queues/${queueUrl}`;
  }
}
