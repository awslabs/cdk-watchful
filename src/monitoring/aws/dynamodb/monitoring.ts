import { GraphWidget, Row } from '@aws-cdk/aws-cloudwatch';
import { IWidget } from '@aws-cdk/aws-cloudwatch/lib/widget';
import { CfnTable, ITable, Operation } from '@aws-cdk/aws-dynamodb';
import { WatchfulContext } from '../../../core/context';
import { WatchfulMetric } from '../../../core/metric';
import { WatchfulMonitoring } from '../../../core/monitoring';
import { SectionWidget } from '../../../widget/section';
import { DynamoDbMetricFactory } from './metrics';

/**
 * Props to create DefaultDynamoDbTableLatencyMonitoring.
 */
export interface DefaultDynamoDbTableMonitoringProps {
  readonly table: ITable;
}

/**
 * Default monitoring for Dynamo DB.
 */
export class DefaultDynamoDbTableMonitoring implements WatchfulMonitoring {
  protected readonly context: WatchfulContext;
  protected readonly metricFactory: DynamoDbMetricFactory;
  protected readonly tableName: string;
  protected readonly provisionedThroughput?: CfnTable.ProvisionedThroughputProperty;
  protected readonly metricProvisionedReadCapacityUnitsMaximum: WatchfulMetric;
  protected readonly metricConsumedReadCapacityUnitsSum: WatchfulMetric;
  protected readonly metricMaxProvisionedReadCapacityUtilization: WatchfulMetric;
  protected readonly metricProvisionedWriteCapacityUnitsMaximum: WatchfulMetric;
  protected readonly metricConsumedWriteCapacityUnitsSum: WatchfulMetric;
  protected readonly metricMaxProvisionedWriteCapacityUtilization: WatchfulMetric;
  protected readonly metricSuccessfulRequestLatencyAverage: WatchfulMetric[];
  protected readonly metricSystemErrorsSum: WatchfulMetric[];

  constructor(context: WatchfulContext, props: DefaultDynamoDbTableMonitoringProps) {
    function enumKeys<O extends object, K extends keyof O = keyof O>(obj: O): K[] {
      return Object.keys(obj).filter(k => Number.isNaN(+k)) as K[];
    }

    this.context = context;
    this.metricFactory = new DynamoDbMetricFactory(context);
    this.tableName = props.table.tableName;
    const cfnTable = props.table.node.defaultChild as CfnTable;
    this.provisionedThroughput = cfnTable.provisionedThroughput as CfnTable.ProvisionedThroughputProperty;
    this.metricSystemErrorsSum = [];
    this.metricSuccessfulRequestLatencyAverage = [];

    for (const key of enumKeys(Operation)) {
      this.metricSystemErrorsSum.push(this.metricFactory.metricSystemErrorsSum(this.tableName, Operation[key]));
      this.metricSuccessfulRequestLatencyAverage.push(this.metricFactory.metricSuccessfulRequestLatencyAverage(this.tableName, Operation[key]));
    }

    const metricProvisionedCapacityUnitsMaximum = this.metricFactory.metricProvisionedCapacityUnitsMaximum(this.tableName);
    this.metricProvisionedReadCapacityUnitsMaximum = metricProvisionedCapacityUnitsMaximum.read;
    this.metricProvisionedWriteCapacityUnitsMaximum = metricProvisionedCapacityUnitsMaximum.write;
    const metricConsumedCapacityUnitsSum = this.metricFactory.metricConsumedCapacityUnitsSum(this.tableName);
    this.metricConsumedReadCapacityUnitsSum = metricConsumedCapacityUnitsSum.read;
    this.metricConsumedWriteCapacityUnitsSum = metricConsumedCapacityUnitsSum.write;
    const metricMaxProvisionedCapacityUtilizationAverage = this.metricFactory.metricMaxProvisionedTableCapacityUtilizationAverage(this.tableName);
    this.metricMaxProvisionedReadCapacityUtilization = metricMaxProvisionedCapacityUtilizationAverage.read;
    this.metricMaxProvisionedWriteCapacityUtilization = metricMaxProvisionedCapacityUtilizationAverage.write;
  }

  widgets(): IWidget[] {
    return [
      new SectionWidget({
        titleMarkdown: `DynamoDB Table **${this.tableName}**`,
        quicklinks: [
          { title: 'Items', url: this.dynamoDbTableItemsUrl() },
          { title: 'Capacity', url: this.dynamoDbTableCapacityUrl() },
        ],
      }),
      new Row(
        this.latencyWidget(6, 5),
        this.readCapacityWidget(6, 5),
        this.writeCapacityWidget(6, 5),
        this.systemErrorWidget(6, 5),
      ),
    ];
  }

  protected dynamoDbTableItemsUrl(): string {
    const awsRegion = this.context.awsAccountRegion;
    return `https://${awsRegion}.console.aws.amazon.com/dynamodbv2/home?region=${awsRegion}#table?name=${this.tableName}&tab=items`;
  }

  protected dynamoDbTableCapacityUrl(): string {
    const awsRegion = this.context.awsAccountRegion;
    return `https://${awsRegion}.console.aws.amazon.com/dynamodbv2/home?region=${awsRegion}#table?name=${this.tableName}&tab=settings`;
  }

  protected latencyWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: 'Latency',
      left: this.metricSuccessfulRequestLatencyAverage.map(metric => metric.unwrap()),
    });
  }

  protected readCapacityWidget(width: number, height: number) {
    if (this.provisionedThroughput && this.provisionedThroughput.readCapacityUnits > 0) {
      return this.readProvisionedCapacityWidget(width, height, this.provisionedThroughput.readCapacityUnits);
    }
    return this.readOnDemandCapacityWidget(width, height);
  }

  protected writeCapacityWidget(width: number, height: number) {
    if (this.provisionedThroughput && this.provisionedThroughput.writeCapacityUnits > 0) {
      return this.writeProvisionedCapacityWidget(width, height, this.provisionedThroughput.writeCapacityUnits);
    }
    return this.writeOnDemandCapacityWidget(width, height);
  }

  protected readProvisionedCapacityWidget(width: number, height: number, capacity: number) {
    return new GraphWidget({
      width,
      height,
      title: 'Read Capacity',
      left: [this.metricConsumedReadCapacityUnitsSum.unwrap()],
      leftAnnotations: [{ label: 'Provisioned', value: capacity }],
      right: [this.metricMaxProvisionedReadCapacityUtilization.unwrap()],
    });
  }

  protected writeProvisionedCapacityWidget(width: number, height: number, capacity: number) {
    return new GraphWidget({
      width,
      height,
      title: 'Write Capacity',
      left: [this.metricConsumedWriteCapacityUnitsSum.unwrap()],
      leftAnnotations: [{ label: 'Provisioned', value: capacity }],
      right: [this.metricMaxProvisionedWriteCapacityUtilization.unwrap()],
    });
  }

  protected readOnDemandCapacityWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: 'Read Capacity',
      left: [
        this.metricProvisionedReadCapacityUnitsMaximum.unwrap(),
        this.metricConsumedReadCapacityUnitsSum.unwrap(),
      ],
      right: [
        this.metricMaxProvisionedReadCapacityUtilization.unwrap(),
      ],
    });
  }

  protected writeOnDemandCapacityWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: 'Write Capacity',
      left: [
        this.metricProvisionedWriteCapacityUnitsMaximum.unwrap(),
        this.metricConsumedWriteCapacityUnitsSum.unwrap(),
      ],
      right: [
        this.metricMaxProvisionedWriteCapacityUtilization.unwrap(),
      ],
    });
  }

  protected systemErrorWidget(width?: number, height?: number) {
    return new GraphWidget({
      width,
      height,
      title: 'System Errors',
      left: this.metricSystemErrorsSum.map(metric => metric.unwrap()),
    });
  }
}
