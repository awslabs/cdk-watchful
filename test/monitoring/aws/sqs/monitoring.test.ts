import { Queue } from '@aws-cdk/aws-sqs';
import { Stack } from '@aws-cdk/core';
import { SqsMonitoring } from '../../../../src/monitoring/aws/sqs/monitoring';

test('snapshot test', () => {
  // GIVEN
  const resources = createTestResources();
  const unitToTest = new SqsMonitoring({ queue: resources.queue });

  // WHEN
  const widgets = unitToTest.getWidgets();
  const resolvedWidgets = resources.stack.resolve(widgets);

  // THEN
  expect(resolvedWidgets).toMatchSnapshot();
});

function createTestResources() {
  const stack = new Stack();
  const queue = new Queue(stack, 'Queue', { queueName: 'DummyQueueName' });
  return { stack, queue };
}
