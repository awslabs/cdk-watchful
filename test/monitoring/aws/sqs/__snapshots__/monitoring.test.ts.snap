// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Array [
  Object {
    "height": 2,
    "markdown": Object {
      "Fn::Join": Array [
        "",
        Array [
          "# SQS Queue **",
          Object {
            "Fn::GetAtt": Array [
              "Queue4A7E3555",
              "QueueName",
            ],
          },
          "**

[button:Overview](https://",
          Object {
            "Ref": "AWS::Region",
          },
          ".console.aws.amazon.com/sqs/v2/home?region=",
          Object {
            "Ref": "AWS::Region",
          },
          "#/queues/",
          Object {
            "Ref": "Queue4A7E3555",
          },
          ")",
        ],
      ],
    },
    "width": 24,
  },
  Object {
    "height": 5,
    "leftMetrics": Array [
      Object {
        "dimensions": Object {
          "QueueName": Object {
            "Fn::GetAtt": Array [
              "Queue4A7E3555",
              "QueueName",
            ],
          },
        },
        "metricName": "ApproximateNumberOfMessagesVisible",
        "namespace": "AWS/SQS",
        "period": Object {
          "amount": 5,
          "unit": Object {
            "inMillis": 60000,
            "isoLabel": "M",
            "label": "minutes",
          },
        },
        "statistic": "Maximum",
      },
      Object {
        "dimensions": Object {
          "QueueName": Object {
            "Fn::GetAtt": Array [
              "Queue4A7E3555",
              "QueueName",
            ],
          },
        },
        "metricName": "NumberOfMessagesSent",
        "namespace": "AWS/SQS",
        "period": Object {
          "amount": 5,
          "unit": Object {
            "inMillis": 60000,
            "isoLabel": "M",
            "label": "minutes",
          },
        },
        "statistic": "Sum",
      },
    ],
    "props": Object {
      "height": 5,
      "left": Array [
        Object {
          "dimensions": Object {
            "QueueName": Object {
              "Fn::GetAtt": Array [
                "Queue4A7E3555",
                "QueueName",
              ],
            },
          },
          "metricName": "ApproximateNumberOfMessagesVisible",
          "namespace": "AWS/SQS",
          "period": Object {
            "amount": 5,
            "unit": Object {
              "inMillis": 60000,
              "isoLabel": "M",
              "label": "minutes",
            },
          },
          "statistic": "Maximum",
        },
        Object {
          "dimensions": Object {
            "QueueName": Object {
              "Fn::GetAtt": Array [
                "Queue4A7E3555",
                "QueueName",
              ],
            },
          },
          "metricName": "NumberOfMessagesSent",
          "namespace": "AWS/SQS",
          "period": Object {
            "amount": 5,
            "unit": Object {
              "inMillis": 60000,
              "isoLabel": "M",
              "label": "minutes",
            },
          },
          "statistic": "Sum",
        },
      ],
      "leftAnnotations": Array [],
      "leftYAxis": Object {
        "min": 0,
        "showUnits": false,
      },
      "title": "Message Count",
      "width": 8,
    },
    "rightMetrics": Array [],
    "width": 8,
  },
  Object {
    "height": 5,
    "leftMetrics": Array [
      Object {
        "dimensions": Object {
          "QueueName": Object {
            "Fn::GetAtt": Array [
              "Queue4A7E3555",
              "QueueName",
            ],
          },
        },
        "metricName": "ApproximateAgeOfOldestMessage",
        "namespace": "AWS/SQS",
        "period": Object {
          "amount": 5,
          "unit": Object {
            "inMillis": 60000,
            "isoLabel": "M",
            "label": "minutes",
          },
        },
        "statistic": "Maximum",
      },
    ],
    "props": Object {
      "height": 5,
      "left": Array [
        Object {
          "dimensions": Object {
            "QueueName": Object {
              "Fn::GetAtt": Array [
                "Queue4A7E3555",
                "QueueName",
              ],
            },
          },
          "metricName": "ApproximateAgeOfOldestMessage",
          "namespace": "AWS/SQS",
          "period": Object {
            "amount": 5,
            "unit": Object {
              "inMillis": 60000,
              "isoLabel": "M",
              "label": "minutes",
            },
          },
          "statistic": "Maximum",
        },
      ],
      "leftAnnotations": Array [],
      "leftYAxis": Object {
        "label": "sec",
        "min": 0,
        "showUnits": false,
      },
      "title": "Oldest Message Age",
      "width": 8,
    },
    "rightMetrics": Array [],
    "width": 8,
  },
  Object {
    "height": 5,
    "leftMetrics": Array [
      Object {
        "dimensions": Object {
          "QueueName": Object {
            "Fn::GetAtt": Array [
              "Queue4A7E3555",
              "QueueName",
            ],
          },
        },
        "metricName": "SentMessageSize",
        "namespace": "AWS/SQS",
        "period": Object {
          "amount": 5,
          "unit": Object {
            "inMillis": 60000,
            "isoLabel": "M",
            "label": "minutes",
          },
        },
        "statistic": "Average",
      },
    ],
    "props": Object {
      "height": 5,
      "left": Array [
        Object {
          "dimensions": Object {
            "QueueName": Object {
              "Fn::GetAtt": Array [
                "Queue4A7E3555",
                "QueueName",
              ],
            },
          },
          "metricName": "SentMessageSize",
          "namespace": "AWS/SQS",
          "period": Object {
            "amount": 5,
            "unit": Object {
              "inMillis": 60000,
              "isoLabel": "M",
              "label": "minutes",
            },
          },
          "statistic": "Average",
        },
      ],
      "leftYAxis": Object {
        "label": "bytes",
        "min": 0,
        "showUnits": false,
      },
      "title": "Message Size",
      "width": 8,
    },
    "rightMetrics": Array [],
    "width": 8,
  },
]
`;
